### PR TITLE
feat: Iceberg V2 delete file support in druid-iceberg-extensions

### PR DESCRIPTION
**Pull request by Shekharrajak**
Ref https://github.com/apache/druid/issues/19190

### Description

In IcebergCatalog.extractSnapshotDataFiles(), line:

```
dataFilePaths.add(task.file().location());
```

discards task.deletes() entirely. Every FileScanTask from tableScan.planFiles() carries a List<DeleteFile> that must
be applied for correct v2 reads. The current code passes only raw file paths to warehouseSource, and Druid's
ParquetReader has zero awareness of Iceberg delete files.

### Changes 

DeleteFileInfo.java               Serializable POJO: path, contentType (POSITION/EQUALITY), equalityFieldIds       
  IcebergFileTaskInputSource.java   Per-task InputSource carrying data file + delete metadata + schema JSON +        
                                    warehouseSource                                                                  
  IcebergNativeRecordReader.java    Manual positional + equality delete application with streaming reads via         
                                    Parquet.read()                                                                   
  IcebergRecordConverter.java       Iceberg Record to Map with full type coverage        

#### Release note

Iceberg V2 Delete File Support: When FileScanTask.deletes() returns non-empty, the extension creates per-task IcebergFileTaskInputSource objects
carrying serializable metadata (data file path, delete file paths/types/equality field IDs, schema JSON). Workers
apply deletes at read time via IcebergNativeRecordReader which reads position-delete and equality-delete Parquet
files, builds filter sets, and streams the data file while skipping deleted rows. V1 tables (no delete files) continue
to use the existing warehouseSource path unchanged.

<hr>

##### Key changed/added classes in this PR
 * `DeleteFileInfo.java`
 * `IcebergFileTaskInputSource.java`
 * `IcebergNativeRecordReader.java`
 * `IcebergRecordConverter.java`
 * `IcebergInputSource.java`
 * `IcebergCatalog.java`
 * IcebergDruidModule.java

<hr>

This PR has:

- [x] been self-reviewed.
   - [x] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added documentation for new or modified features or behaviors.
- [x] a release note entry in the PR description.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: extensions-contrib/druid-iceberg-extensions/src/main/java/org/apache/druid/iceberg/input/IcebergNativeRecordReader.java [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+    final Schema tableSchema = SchemaParser.fromJson(tableSchemaJson);
+
+    // Step 1: Collect positional deletes
+    final Set<Long> deletedPositions = collectPositionalDeletes();
+
+    // Step 2: Collect equality deletes
+    final List<EqualityDeleteSet> equalityDeleteSets = collectEqualityDeletes(tableSchema);
+
+    // Step 3: Stream data file with delete application
+    final InputFile dataInputFile = HadoopInputFile.fromLocation(dataFilePath, hadoopConf);
```
Comment (FrankChen021) [inline extensions-contrib/druid-iceberg-extensions/src/main/java/org/apache/druid/iceberg/input/IcebergNativeRecordReader.java]: [P1] Native reader bypasses configured warehouse access

The v2 path reads data and delete files with HadoopInputFile.fromLocation(..., new Configuration()) and never uses the configured warehouseSource or catalog FileIO properties. Existing specs that use s3/gcs/azure warehouse sources, dynamic credentials, or non-default Hadoop configuration will work on the v1 path but fail as soon as delete files are present because workers try to open the object-store paths with an empty Hadoop configuration.

**Review by FrankChen021 — COMMENTED** [review]
Findings that could not be attached inline:
- extensions-contrib/druid-iceberg-extensions/src/main/java/org/apache/druid/iceberg/input/IcebergInputSource.java:164 - [P1] V2 tables with deletes produce zero splits. When any delete file is present, retrieveIcebergDatafiles() sets delegateInputSource to EmptyInputSource, so createSplits() returns an empty stream and estimateNumSplits() returns 0. MSQ and parallel ingestion slice SplittableInputSource inputs exclusively through createSplits()/withSplit(), so an Iceberg v2 table with deletes will schedule no readable input slices and ingest no rows instead of using the native reader path.

File: extensions-contrib/druid-iceberg-extensions/src/main/java/org/apache/druid/iceberg/input/IcebergNativeRecordReader.java (hunk 1 above)
Comment (Shekharrajak) [reply extensions-contrib/druid-iceberg-extensions/src/main/java/org/apache/druid/iceberg/input/IcebergNativeRecordReader.java]: Thanks for checking - updated : https://github.com/apache/druid/pull/19266/changes/23ef4dbe6c910b238c0379943b58f7fe2d954d51#diff-1b9776e43fa17c32a610eee043a6fd6cbf32b29ae4026d24ea798ac9a6f638bcR299

**Comment by Shekharrajak** [comment]
> * extensions-contrib/druid-iceberg-extensions/src/main/java/org/apache/druid/iceberg/input/IcebergInputSource.java:164 - [P1] V2 tables with deletes produce zero splits. When any delete file is present, retrieveIcebergDatafiles() sets delegateInputSource to EmptyInputSource, so createSplits() returns an empty stream and estimateNumSplits() returns 0. MSQ and parallel ingestion slice SplittableInputSource inputs exclusively through createSplits()/withSplit(), so an Iceberg v2 table with deletes will schedule no readable input slices and ingest no rows instead of using the native reader path.

updated : https://github.com/apache/druid/pull/19266/changes/309c97b387f9a7012c218ae35659654e7c13f0e3#diff-1b9776e43fa17c32a610eee043a6fd6cbf32b29ae4026d24ea798ac9a6f638bcR168

**Comment by Shekharrajak** [comment]
Noted gaps in iceberg v2 spec support https://github.com/apache/druid/issues/19471

**Review by FrankChen021 — COMMENTED** [review]
| Severity | Findings |
| --- | ---: |
| P0 | 0 |
| P1 | 0 |
| P2 | 1 |
| P3 | 0 |
| Total | 1 |

| Severity | Findings |
| --- | ---: |
| P0 | 0 |
| P1 | 0 |
| P2 | 1 |
| P3 | 0 |
| Total | 1 |

Reviewed 15 of 15 changed files. The earlier FileIO metadata follow-up is addressed, so no inline reply is needed; the new finding is below.

---

This is an automated review by Codex GPT-5.5

File: extensions-contrib/druid-iceberg-extensions/src/main/java/org/apache/druid/iceberg/input/IcebergNativeRecordReader.java [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+
+    // Step 1: Collect positional deletes
+    final Set<Long> deletedPositions = collectPositionalDeletes();
+
+    // Step 2: Collect equality deletes
+    final List<EqualityDeleteSet> equalityDeleteSets = collectEqualityDeletes(tableSchema);
+
+    // Step 3: Stream data file with delete application
+    final InputFile dataInputFile = fileIO.newInputFile(dataFilePath);
+    final CloseableIterable<Record> records = Parquet.read(dataInputFile)
```
Comment (FrankChen021) [inline extensions-contrib/druid-iceberg-extensions/src/main/java/org/apache/druid/iceberg/input/IcebergNativeRecordReader.java:156]: [P2] V2 delete path only reads Parquet files

The new v2 path bypasses the configured inputFormat and always opens the data file with Parquet.read(...), and the delete readers do the same for delete files. The extension still documents Iceberg data files as Parquet, ORC, or Avro via the delegated warehouseSource/inputFormat path, so any ORC/Avro Iceberg table with delete files will now route through this native reader and fail to parse the files instead of using the configured format-specific reader. Please either dispatch by Iceberg FileFormat / reader support or explicitly reject/document v2 delete support as Parquet-only.

Comment (Shekharrajak) [reply extensions-contrib/druid-iceberg-extensions/src/main/java/org/apache/druid/iceberg/input/IcebergNativeRecordReader.java:156]: Let me document for parquet only and follow up PR will add it : https://github.com/apache/druid/issues/19472

Comment (FrankChen021) [reply extensions-contrib/druid-iceberg-extensions/src/main/java/org/apache/druid/iceberg/input/IcebergNativeRecordReader.java:156]: Thanks, this now guards the data file format, but I think one edge of the original concern is still open: Iceberg delete files carry their own `deleteFile.format()`, separate from `task.file().format()`. Right now `DeleteFileInfo` only carries path/content/equality fields, and `requireParquet(deleteFileInfo.getPath())` checks the data file format, so a PARQUET data file with an ORC/AVRO delete file would still reach `Parquet.read(deleteInputFile)` instead of being explicitly rejected. Could we preserve the delete-file format and guard each delete file too?

Also, the user docs still say Iceberg data files can be Parquet, ORC, or Avro and the new v2 delete section does not call out the Parquet-only limitation, so it would be worth documenting that limitation there as well.

**Comment by Shekharrajak** [comment]
Flaky test reported https://github.com/apache/druid/issues/19491 

Please help in triggering the one failed CI check run. We can work on this flaky test separately.

**Review by FrankChen021 — COMMENTED** [review]
I have reviewed the code for correctness, edge cases, concurrency, and integration risks; no issues found.

Reviewed 15 of 15 changed files.

---

This is an automated review by Codex GPT-5.5

File: extensions-contrib/druid-iceberg-extensions/pom.xml [hunk 3]
```diff
@@ -760,17 +760,85 @@
       <scope>provided</scope>
     </dependency>
 
+    <dependency>
+      <groupId>org.apache.iceberg</groupId>
+      <artifactId>iceberg-data</artifactId>
+      <version>${iceberg.core.version}</version>
+      <scope>runtime</scope>
```
Comment (FrankChen021) [inline extensions-contrib/druid-iceberg-extensions/pom.xml]: [P1] iceberg-data is not on the main compile classpath

The new production classes import `org.apache.iceberg.data.Record` and `GenericParquetReaders`, but `iceberg-data` is declared with runtime scope. Maven does not put runtime-scoped dependencies on the main compile classpath, so this module cannot compile once `IcebergNativeRecordReader` and `IcebergRecordConverter` are compiled. Use the default compile scope for `iceberg-data`.

File: extensions-contrib/druid-iceberg-extensions/src/main/java/org/apache/druid/iceberg/input/IcebergNativeRecordReader.java [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+      @Nullable final Map<String, String> fileIOProperties,
+      final String fileFormat
+  )
+  {
+    this.dataFilePath = dataFilePath;
+    this.deleteFiles = deleteFiles;
+    this.tableSchemaJson = tableSchemaJson;
+    this.warehouseSource = warehouseSource;
+    this.inputRowSchema = inputRowSchema;
+    this.hadoopConf = new Configuration();
```
Comment (FrankChen021) [inline extensions-contrib/druid-iceberg-extensions/src/main/java/org/apache/druid/iceberg/input/IcebergNativeRecordReader.java]: [P2] V2 reader drops configured warehouse/Hadoop access

The v2 path constructs a fresh Hadoop `Configuration` and then opens data and delete files through a recreated Iceberg `FileIO`, while the configured `warehouseSource` is only stored and never used. Existing HDFS/S3 warehouse ingestion relies on `warehouseSource` and injected Hadoop configuration, so v2 tables with deletes can plan successfully on the controller but fail on workers when those files require the Druid warehouse source or cluster Hadoop credentials.

**Review by FrankChen021 — COMMENTED** [review]
| Severity | Findings |
| --- | ---: |
| P0 | 0 |
| P1 | 1 |
| P2 | 1 |
| P3 | 0 |
| Total | 2 |

Reviewed 15 of 15 changed files.

---

This is an automated review by Codex GPT-5.5

File: extensions-contrib/druid-iceberg-extensions/pom.xml (hunk 3 above)
Comment (Shekharrajak) [reply extensions-contrib/druid-iceberg-extensions/pom.xml]: https://github.com/apache/druid/pull/19266/changes/5ecc28cfb2d8ef114b40eac2de35cb00d1dd03d7

Comment (Shekharrajak) [reply extensions-contrib/druid-iceberg-extensions/pom.xml]: updated

File: extensions-contrib/druid-iceberg-extensions/src/main/java/org/apache/druid/iceberg/input/IcebergFileTaskInputSource.java [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+  @JsonCreator
+  public IcebergFileTaskInputSource(
+      @JsonProperty("dataFilePath") final String dataFilePath,
+      @JsonProperty("deleteFiles") final List<DeleteFileInfo> deleteFiles,
+      @JsonProperty("tableSchemaJson") final String tableSchemaJson,
+      @JsonProperty("warehouseSource") final InputSourceFactory warehouseSource,
+      @JsonProperty("fileIOImpl") @Nullable final String fileIOImpl,
+      @JsonProperty("fileIOProperties") @Nullable final Map<String, String> fileIOProperties,
+      @JsonProperty("fileFormat") @Nullable final String fileFormat,
+      final Configuration hadoopConf
```
Comment (FrankChen021) [inline extensions-contrib/druid-iceberg-extensions/src/main/java/org/apache/druid/iceberg/input/IcebergFileTaskInputSource.java]: [P1] Unbound Configuration breaks v2 split deserialization

IcebergFileTaskInputSource is registered as a Jackson subtype and returned from withSplit for v2 tables, so workers need to deserialize it. The @JsonCreator leaves hadoopConf as an unannotated constructor parameter, unlike the catalog classes that use @JacksonInject @HiveConf, so Druid's ObjectMapper has no property or injectable value to bind. This can make v2 delete splits fail before reading. Inject it, annotate it, or create a safe default, and add a Jackson round-trip test for this input source.

File: extensions-contrib/druid-iceberg-extensions/src/main/java/org/apache/druid/iceberg/input/IcebergNativeRecordReader.java [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+
+    // Step 1: Collect positional deletes
+    final Set<Long> deletedPositions = collectPositionalDeletes();
+
+    // Step 2: Collect equality deletes
+    final List<EqualityDeleteSet> equalityDeleteSets = collectEqualityDeletes(tableSchema);
+
+    // Step 3: Stream data file with delete application
+    requireParquet(dataFilePath);
+    final InputFile dataInputFile = fileIO.newInputFile(dataFilePath);
```
Comment (FrankChen021) [inline extensions-contrib/druid-iceberg-extensions/src/main/java/org/apache/druid/iceberg/input/IcebergNativeRecordReader.java:155]: [P1] V2 path bypasses warehouseSource file access

When delete files are present, the new reader opens data and delete files through Iceberg FileIO instead of the configured warehouseSource/inputFormat. Existing Iceberg specs use warehouseSource for S3/GCS/local access settings, endpoints, and credentials, so a table that only adds delete files can silently switch file-access mechanisms and fail or read from the wrong filesystem. The v2 path should preserve or translate the warehouseSource-backed access configuration for native reads.

**Review by FrankChen021 — COMMENTED** [review]
| Severity | Findings |
| --- | ---: |
| P0 | 0 |
| P1 | 2 |
| P2 | 0 |
| P3 | 0 |
| Total | 2 |

The original `iceberg-data` dependency follow-up is handled; I found two remaining v2 read-path issues.

Reviewed 16 of 16 changed files.

---

This is an automated review by Codex GPT-5.5

**Review by FrankChen021 — COMMENTED** [review]
I reviewed the incremental update and full changed-file set for correctness, edge cases, concurrency, and integration risks; no issues found.

Reviewed 17 of 17 changed files.

---

This is an automated review by Codex GPT-5.5

**Comment by jtuglu1** [comment]
Related: https://github.com/apache/druid/pull/19534

**Review by FrankChen021 — COMMENTED** [review]
I reviewed the incremental update and full changed-file set for correctness, edge cases, lifecycle, API compatibility, and integration risks; no issues found.

Reviewed 17 of 17 changed files.

---

This is an automated review by Codex GPT-5.5
